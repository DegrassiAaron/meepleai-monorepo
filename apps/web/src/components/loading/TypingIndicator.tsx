/**
 * TypingIndicator Component
 *
 * Animated typing indicator for chat interfaces.
 * Shows 3 bouncing dots with staggered animation.
 * Respects user's reduced motion preferences.
 *
 * @example
 * ```tsx
 * <TypingIndicator
 *   visible={isTyping}
 *   agentName="AI Assistant"
 * />
 * ```
 */

import { motion, AnimatePresence } from 'framer-motion';
import { useReducedMotion } from '@/lib/animations';

export interface TypingIndicatorProps {
  /**
   * Whether the indicator is visible
   */
  visible: boolean;

  /**
   * Name of the agent that is typing (for accessibility)
   */
  agentName: string;

  /**
   * Additional CSS classes
   */
  className?: string;
}

/**
 * TypingIndicator component for chat interfaces
 */
export function TypingIndicator({
  visible,
  agentName,
  className = '',
}: TypingIndicatorProps) {
  const shouldReduceMotion = useReducedMotion();

  // Animation variants for container
  const containerVariants = {
    initial: { opacity: 0, scale: 0.95 },
    animate: { opacity: 1, scale: 1 },
    exit: { opacity: 0, scale: 0.95 },
  };

  // Animation variants for dots (bounce effect with stagger)
  const dotVariants = shouldReduceMotion
    ? {
        initial: { y: 0 },
        animate: { y: 0 },
      }
    : {
        initial: { y: 0 },
        animate: {
          y: [-2, 0, -2],
        },
      };

  // Transition configuration for dots
  const getDotTransition = (delay: number) =>
    shouldReduceMotion
      ? { duration: 0 }
      : {
          duration: 0.6,
          repeat: Infinity,
          ease: 'easeInOut' as const,
          delay,
        };

  // Stagger delays for each dot
  const staggerDelays = [0, 0.1, 0.2];

  return (
    <AnimatePresence>
      {visible && (
        <motion.div
          role="status"
          aria-live="polite"
          aria-label={`${agentName} is typing`}
          variants={containerVariants}
          initial="initial"
          animate="animate"
          exit="exit"
          transition={{ duration: 0.2 }}
          className={`flex items-center space-x-1 px-4 py-2 bg-slate-100 dark:bg-slate-800 rounded-2xl w-fit ${className}`}
        >
          {/* Screen reader text */}
          <span className="sr-only">{agentName} is typing...</span>

          {/* Animated dots */}
          <div className="flex items-center space-x-1" aria-hidden="true">
            {staggerDelays.map((delay, index) => (
              <motion.span
                key={index}
                variants={dotVariants}
                initial="initial"
                animate="animate"
                transition={getDotTransition(delay)}
                className="w-2 h-2 bg-slate-500 dark:bg-slate-400 rounded-full"
              />
            ))}
          </div>
        </motion.div>
      )}
    </AnimatePresence>
  );
}
