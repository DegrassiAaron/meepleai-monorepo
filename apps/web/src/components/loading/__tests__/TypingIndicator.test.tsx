/**
 * Tests for TypingIndicator component
 * Animated typing indicator for chat interfaces
 */

import { render, screen } from '@testing-library/react';
import { TypingIndicator } from '../TypingIndicator';

// Mock useReducedMotion hook
const mockUseReducedMotion = jest.fn();
jest.mock('@/lib/animations', () => ({
  ...jest.requireActual('@/lib/animations'),
  useReducedMotion: () => mockUseReducedMotion(),
}));

// Mock framer-motion to avoid animation issues in tests
jest.mock('framer-motion', () => ({
  motion: {
    div: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    span: ({ children, ...props }: any) => <span {...props}>{children}</span>,
  },
  AnimatePresence: ({ children }: any) => <>{children}</>,
}));

describe('TypingIndicator', () => {
  beforeEach(() => {
    mockUseReducedMotion.mockReturnValue(false);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Visibility', () => {
    it('should render when visible prop is true', () => {
      render(<TypingIndicator visible={true} agentName="AI Assistant" />);
      expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('should not render when visible prop is false', () => {
      render(<TypingIndicator visible={false} agentName="AI Assistant" />);
      expect(screen.queryByRole('status')).not.toBeInTheDocument();
    });
  });

  describe('Agent name display', () => {
    it('should display agentName in accessible text', () => {
      render(<TypingIndicator visible={true} agentName="AI Assistant" />);
      const status = screen.getByRole('status');
      expect(status).toHaveAttribute('aria-label', 'AI Assistant is typing');
    });

    it('should handle different agent names', () => {
      render(<TypingIndicator visible={true} agentName="Chess Expert" />);
      const status = screen.getByRole('status');
      expect(status).toHaveAttribute('aria-label', 'Chess Expert is typing');
    });

    it('should display screen reader text with agent name', () => {
      render(<TypingIndicator visible={true} agentName="Support Bot" />);
      expect(screen.getByText('Support Bot is typing...')).toBeInTheDocument();
    });
  });

  describe('Dots rendering', () => {
    it('should render 3 dots', () => {
      const { container } = render(
        <TypingIndicator visible={true} agentName="AI" />
      );
      const dots = container.querySelectorAll('.w-2.h-2');
      expect(dots).toHaveLength(3);
    });

    it('should render dots with rounded-full class', () => {
      const { container } = render(
        <TypingIndicator visible={true} agentName="AI" />
      );
      const dots = container.querySelectorAll('.rounded-full');
      expect(dots.length).toBeGreaterThanOrEqual(3);
    });
  });

  describe('Reduced motion', () => {
    it('should respect prefers-reduced-motion', () => {
      mockUseReducedMotion.mockReturnValue(true);
      const { container } = render(
        <TypingIndicator visible={true} agentName="AI" />
      );
      const status = container.querySelector('[role="status"]');
      expect(status).toBeInTheDocument();
      // Component should still render but with no animation variants
      // Animation behavior is mocked in tests, so we just verify it renders
    });

    it('should render with animation when reduced motion is false', () => {
      mockUseReducedMotion.mockReturnValue(false);
      render(<TypingIndicator visible={true} agentName="AI" />);
      expect(screen.getByRole('status')).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('should have correct ARIA attributes', () => {
      render(<TypingIndicator visible={true} agentName="AI Assistant" />);
      const status = screen.getByRole('status');
      expect(status).toHaveAttribute('role', 'status');
      expect(status).toHaveAttribute('aria-live', 'polite');
      expect(status).toHaveAttribute('aria-label', 'AI Assistant is typing');
    });

    it('should include screen reader text', () => {
      render(<TypingIndicator visible={true} agentName="Helper Bot" />);
      const srText = screen.getByText('Helper Bot is typing...');
      expect(srText).toHaveClass('sr-only');
    });
  });

  describe('Custom className', () => {
    it('should apply custom className', () => {
      const { container } = render(
        <TypingIndicator
          visible={true}
          agentName="AI"
          className="custom-typing-indicator"
        />
      );
      const wrapper = container.querySelector('[role="status"]');
      expect(wrapper).toHaveClass('custom-typing-indicator');
    });

    it('should preserve base classes with custom className', () => {
      const { container } = render(
        <TypingIndicator visible={true} agentName="AI" className="mt-4" />
      );
      const wrapper = container.querySelector('[role="status"]');
      expect(wrapper).toHaveClass('mt-4');
      expect(wrapper).toHaveClass('flex'); // Base class
    });
  });

  describe('Snapshot test', () => {
    it('should match snapshot when visible', () => {
      const { container } = render(
        <TypingIndicator visible={true} agentName="AI Assistant" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot when not visible', () => {
      const { container } = render(
        <TypingIndicator visible={false} agentName="AI Assistant" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
