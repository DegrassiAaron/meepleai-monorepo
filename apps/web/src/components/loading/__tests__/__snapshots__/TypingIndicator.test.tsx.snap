// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TypingIndicator Snapshot test should match snapshot when not visible 1`] = `null`;

exports[`TypingIndicator Snapshot test should match snapshot when visible 1`] = `
<div
  animate="animate"
  aria-label="AI Assistant is typing"
  aria-live="polite"
  class="flex items-center space-x-1 px-4 py-2 bg-slate-100 dark:bg-slate-800 rounded-2xl w-fit "
  exit="exit"
  initial="initial"
  role="status"
  transition="[object Object]"
  variants="[object Object]"
>
  <span
    class="sr-only"
  >
    AI Assistant
     is typing...
  </span>
  <div
    aria-hidden="true"
    class="flex items-center space-x-1"
  >
    <span
      animate="animate"
      class="w-2 h-2 bg-slate-500 dark:bg-slate-400 rounded-full"
      initial="initial"
      transition="[object Object]"
      variants="[object Object]"
    />
    <span
      animate="animate"
      class="w-2 h-2 bg-slate-500 dark:bg-slate-400 rounded-full"
      initial="initial"
      transition="[object Object]"
      variants="[object Object]"
    />
    <span
      animate="animate"
      class="w-2 h-2 bg-slate-500 dark:bg-slate-400 rounded-full"
      initial="initial"
      transition="[object Object]"
      variants="[object Object]"
    />
  </div>
</div>
`;
